presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1411,48 +1414,82 @@ else if (expression instanceof LongLiteral) {
         return orderByExpressions;
     }
 
-    private List<FieldOrExpression> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
+    private List<List<FieldOrExpression>> analyzeGroupBy(QuerySpecification node, RelationType tupleDescriptor, AnalysisContext context, List<FieldOrExpression> outputExpressions)
     {
-        ImmutableList.Builder<FieldOrExpression> groupByExpressionsBuilder = ImmutableList.builder();
-        if (!node.getGroupBy().isEmpty()) {
-            // Translate group by expressions that reference ordinals
-            for (Expression expression : node.getGroupBy()) {
-                // first, see if this is an ordinal
-                FieldOrExpression groupByExpression;
+        if (node.getGroupBy().isEmpty()) {
+            analysis.setGroupingSets(node, ImmutableList.of());
+            return ImmutableList.of();
+        }
 
-                if (expression instanceof LongLiteral) {
-                    long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > outputExpressions.size()) {
-                        throw new SemanticException(INVALID_ORDINAL, expression, "GROUP BY position %s is not in select list", ordinal);
-                    }
+        ImmutableList.Builder<List<FieldOrExpression>> groupingSetsBuilder = ImmutableList.builder();
+        List<GroupingElement> groupingElements = node.getGroupBy().stream()
```

Comment:
Well, the cross product happens after all the SimplyGroupBy's are merged. If we had `GROUP BY (a, b), (c, d), ROLLUP (e, f)`, this should result in the equivalence of
`GROUPING SETS ((a, b, c, d, e, f), (a, b, c, d, e), (a, b, c, d, f), (a, b, c, d))`.